scio-core/src/main/scala/com/spotify/scio/coders/KryoAtomicCoder.scala

```diff
@@ -76,168 +80,175 @@ private object KryoRegistrarLoader {
-            s"elapsed: $elapsed ms, count: $count, bytes: $bytes")
         }
-      }
-    case _ =>
-      observer.update(kryoEncodedElementByteSize(value))
-  }
+      case _ =>
+        observer.update(kryoEncodedElementByteSize(value))
+    }
 
-  private def kryoEncodedElementByteSize(obj: Any): Long = {
+  private def kryoEncodedElementByteSize(obj: Any): Long = kryoPool(options) { kryo =>
     val s = new CountingOutputStream(ByteStreams.nullOutputStream())
     val output = new Output(options.bufferSize, options.maxBufferSize)
     output.setOutputStream(s)
-    kryoState.get().kryo.writeClassAndObject(output, obj)
+    kryo.writeClassAndObject(output, obj)
     output.flush()
     s.getCount + VarInt.getLength(s.getCount)
   }
 
 }
 
 /** Used for sharing Kryo instance and buffers. */
-private[scio] final case class KryoState(kryo: Kryo, input: InputChunked, output: OutputChunked)
+private[scio] final case class KryoState(input: InputChunked, output: OutputChunked)
 
 private[scio] object KryoAtomicCoder {
   private val logger = LoggerFactory.getLogger(this.getClass)
+
+  private[this] val kryoPool: KryoPool = {
+    val factory = new KryoFactory {
+      override def create(): Kryo = {
+        val k = KryoSerializer.registered.newKryo()
+        k.forClass(new CoderSerializer(InstantCoder.of()))
+        k.forClass(new CoderSerializer(TableRowJsonCoder.of()))
+        // Java Iterable/Collection are missing proper equality check, use custom CBF as a
+        // workaround
+        k.register(classOf[Wrappers.JIterableWrapper[_]],
+                   new JTraversableSerializer[Any, Iterable[Any]]()(new JIterableWrapperCBF[Any]))
+        k.register(classOf[Wrappers.JCollectionWrapper[_]],
+                   new JTraversableSerializer[Any, Iterable[Any]]()(new JCollectionWrapperCBF[Any]))
+        // Wrapped Java collections may have immutable implementations, i.e. Guava, treat them
+        // as regular Scala collections as a workaround
+        k.register(classOf[Wrappers.JListWrapper[_]],
+                   new JTraversableSerializer[Any, mutable.Buffer[Any]])
+        k.forSubclass[SpecificRecordBase](new SpecificAvroSerializer)
+        k.forSubclass[GenericRecord](new GenericAvroSerializer)
+        k.forSubclass[Message](new ProtobufSerializer)
+        k.forSubclass[LocalDate](new JodaLocalDateSerializer)
+        k.forSubclass[LocalTime](new JodaLocalTimeSerializer)
+        k.forSubclass[LocalDateTime](new JodaLocalDateTimeSerializer)
+        k.forSubclass[DateTime](new JodaDateTimeSerializer)
+        k.forSubclass[Path](new JPathSerializer)
+        k.forSubclass[ByteString](new ByteStringSerializer)
+        k.forClass(new KVSerializer)
+        // TODO:
+        // TimestampedValueCoder
+        new AlgebirdRegistrar()(k)
+        KryoRegistrarLoader.load(k)
+
+        k
+      }
+    }
+
+    new KryoPool.Builder(factory).softReferences().build()
+  }
+
+  def kryoPool[R](options: KryoOptions)(f: Kryo => R): R = {
```

Comment:
Nitpick: the name of this method is isn't the best, right? `kryoPool` - what does that mean? Am I creating a KryoPool, am I initializing it, am I getting an instance? Maybe `withKryo`, `borrowKryo` or maybe sth different? I like `withKryo` tho.

Reply:
ok I can change it to `withKryo` no strong feelings